johnpooch/diplomacy
# Add turn end functionality

Implements turn end functionality. This basically means that a turn can have a datetime in the future at which the turn will be processed and the game will be advanced.

* Adds `TurnEndManager` which provides a `new` method which should be used for creating `TurnEnd` instances instead of `create`. 
* Adds a `turn_end` property to `Turn` which gets the related `TurnEnd` instance or `None`.
* Adds `set_turn_end` management command which can be used to manually set the `TurnEnd` of a game.
* Adds functionality to automatically process the game at turn end datetime
* Exposes the `TurnEnd.datetime` through the API so we can add a timer to the client.

**How to test**
Bring up containers and run `make` to build fixtures. Now set the turn end of a game to one minute from now.
`./manage.py set_turn_end 'first-turn' '2020-11-08 19:55:30'` (update the datetime!). Stare at the screen for a minute. If you've got the logs up you'll see worker container spit out some logs. Check the game and you'll see that the turn has been successfully processed.

**Note**
The game/turn processing logic makes me nervous. I think it needs a lot more testing until we can be sure that it behaves as we expect it to. Not relevant for this PR though.
